File: algorithms/background/gmodel/modeller.py
#!/usr/bin/env python
#
# algorithm.py
#
#  Copyright (C) 2013 Diamond Light Source
#
#  Author: James Parkhurst
#
#  This code is distributed under the BSD license, a copy of which is
#  included in the root directory of this package.

from __future__ import division

from dials.phil import parse

phil_scope = parse('''

  min_count = 5
    .type = int(value_min=1)

  nsigma = 6
    .type = float(value_min=0)

  sigma = 0.5
    .type = float(value_min=0)

  kernel_size = 9
    .type = int

  niter = 10
    .type = int

''')

class Modeller(object):

  def __init__(self, beam, detector,
               min_count=5, nsigma=6, sigma=0.5,
               kernel_size=9, niter=10):
    from dials.algorithms.background.gmodel import PixelFilter

    self.beam = beam
    self.detector = detector

    self.min_count = min_count
    self.nsigma = nsigma
    self.sigma = sigma
    self.kernel_size = kernel_size
    self.niter = niter

    width, height = detector[0].get_image_size()

    self._filter = PixelFilter(height, width)

    self.detector_mask = None

  def add_image(self, frame, image, mask, reflections):

    height, width = image.all()

    _,_,_,_,z0,z1 = reflections['bbox'].parts()
    selection = (z0 <= frame) & (z1 > frame)
    subset = reflections.select(selection)
    sbox_mask = subset['shoebox'].apply_background_mask(frame, 1, (height, width))

    if self.detector_mask is None:
      self.detector_mask = mask

    mask = mask & sbox_mask

    self._filter.add(image, mask)

  def compute(self):
    from dials.algorithms.background.gmodel import FillGaps
    from dials.algorithms.image.fill_holes import simple_fill
    result = self._filter.compute(self.min_count, self.nsigma)

    data = result.data()
    mask = result.mask()

    data = simple_fill(data, mask)

    fill = FillGaps(self.beam, self.detector[0])

    mask = mask.as_1d().as_int()
    mask = mask - (~self.detector_mask).as_1d().as_int()
    mask.reshape(data.accessor())

    fill(data, mask, self.sigma, self.kernel_size, self.niter)

    return data


class Creator(object):

  def __init__(self, experiment, params):

    self.modeller = None
    self.background = None
    self.experiment = experiment
    self.params = params

  def initialized(self):
    return self.modeller is not None

  def finalized(self):
    return self.background is not None

  def initialize(self):
    assert not self.initialized()
    self.modeller = Modeller(
      self.experiment.beam,
      self.experiment.detector,
      self.params.min_count,
      self.params.nsigma)

  def next_image(self, frame, image, mask, reflections):
    assert self.initialized()
    self.modeller.add_image(frame, image, mask, reflections)

  def finalize(self):
    assert self.initialized()
    self.background = self.modeller.compute()
    self.modeller = None
    if self.params.debug.output:
      import cPickle as pickle
      from logging import info
      filename = self.params.debug.filename
      info("Writing background model to %s" % filename)
      with open(filename, "w") as outfile:
        pickle.dump(self.background, outfile, protocol=pickle.HIGHEST_PROTOCOL)

  def compute(self, reflections):
    from dials.algorithms.background.gmodel import Fitter
    from dials.array_family import flex
    assert self.finalized()
    fitter = Fitter(self.background)
    scale = fitter(reflections['shoebox'])
    success = scale >= 0
    mean = flex.double([
      flex.mean(sbox.background)
      for sbox in reflections['shoebox']
    ])
    reflections['background.mean'] = mean
    reflections['background.scale'] = scale
    reflections.set_flags(success != True, reflections.flags.dont_integrate)
    return success

File: algorithms/background/simple/boost_python/modeller.cc
/*
 * modeller.cc
 *
 *  Copyright (C) 2013 Diamond Light Source
 *
 *  Author: James Parkhurst
 *
 *  This code is distributed under the BSD license, a copy of which is
 *  included in the root directory of this package.
 */
#include <boost/python.hpp>
#include <boost/python/def.hpp>
#include <dials/algorithms/background/simple/modeller.h>

namespace dials { namespace algorithms { namespace background { namespace boost_python {

  using namespace boost::python;

  void export_modeller() {
    // An abstract class for background model
    class_<Model, boost::shared_ptr<Model>, boost::noncopyable>("Model", no_init)
      .def("value", &Model::value, (arg("z"), arg("y"), arg("x")))
      .def("params", &Model::params)
      .def("variances", &Model::variances);

    // An abtract class for background modeller
    class_<Modeller, boost::noncopyable>("Modeller", no_init)
      .def("create", &Modeller::create, (arg("data"), arg("mask")));

    // A constant 2d model
    class_<Constant2dModel, bases<Model> >("Constant2dModel", no_init)
      .def(init<af::shared<double>, af::shared<double> >((arg("a"), arg("va"))));

    // A constant 3d model
    class_<Constant3dModel, bases<Model> >("Constant3dModel", no_init)
      .def(init<double, double>((arg("a"), arg("va"))));

    // A linear 2d model
    class_<Linear2dModel, bases<Model> >("Linear2dModel", no_init)
      .def(init<af::shared<double>,
                af::shared<double>,
                af::shared<double>,
                af::shared<double>,
                af::shared<double>,
                af::shared<double> >(
        (arg("a"), arg("b"), arg("c"), arg("va"), arg("vb"), arg("vc"))));

    // A linear 3d model
    class_<Linear3dModel, bases<Model> >("Linear3dModel", no_init)
      .def(init<double, double, double, double, double, double, double, double>(
        (arg("a"),
         arg("b"),
         arg("c"),
         arg("d"),
         arg("va"),
         arg("vb"),
         arg("vc"),
         arg("vd"))));

    // The modeller classes
    class_<Constant2dModeller, bases<Modeller> >("Constant2dModeller");
    class_<Constant3dModeller, bases<Modeller> >("Constant3dModeller");
    class_<Linear2dModeller, bases<Modeller> >("Linear2dModeller");
    class_<Linear3dModeller, bases<Modeller> >("Linear3dModeller");
  }

}}}}  // namespace dials::algorithms::background::boost_python
